types/index.d.ts

```diff
@@ -0,0 +1,116 @@
+import { ObjectSchema, Schema } from 'yup'
+import { ReactElement, ReactNode, SyntheticEvent, ReactType } from 'react'
+import { RenderedAsComponent } from './helpers'
+
+type Errors = Record<string, any[]>
+type Touched = Record<string, boolean>
+
+export type SchemaValue<T extends ObjectSchema<unknown>> = ReturnType<T['cast']>
+
+type ValidateData = {
+  fields: string[]
+  type: SyntheticEvent
+  args?: any[]
+}
+
+type BeforeSubmitData<T> = {
+  value: T
+  errors: Errors
+}
+
+type Mapper<TOut, TIn = any> = (input: TIn) => TOut
+
+type Meta<TValue = any, TSchema extends Schema = Schema<TValue>> = {
+  schema: TSchema | undefined
+  errors: Errors
+  touched: Touched
+  errorClass: string
+  context?: {}
+  invalid: boolean
+  valid: boolean
+  value: TValue
+  onError(Errors): void
+}
+
+interface FieldRenderProps<TValue = any> {
+  type: string
+  meta: Meta<TValue>
+}
+
+interface FieldProps<TValue = any> {
+  name: string
+  events?: string | string[] | ((meta: Meta<TValue>) => string | string[])
+  mapToValue?: Mapper<any, TValue>
+  mapFromValue?:
+    | Mapper<TValue>
+    | keyof Value
+    | { [P in keyof TValue]: string | Mapper<TValue> }
+  exclusive?: boolean
+  noValidate?: boolean
+  errorClass?: string
+  validates?: string | string[]
+  noMeta?: boolean
+
+  children?: ReactNode | ((opts: FieldRenderProps<TValue>) => ReactNode)
+}
+
+class FormField<
+  TValue,
+  As extends ReactType = 'input'
```

Comment:
yeah I am not certain that type inference will work the way you want here... If it works, great! otherwise I think adding `[idx: string]: any` in the props is a safer bet, or maybe explicitly pass that as a generic